File: Data Structures & Algorithms/balanced-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        res=0
        node=True
        def height(curr):
            nonlocal node
            if curr==None:
                return 0
            left=height(curr.left)
            right=height(curr.right)
            if abs(left-right)>1:
                 node=False
            return 1+max(left,right)
        height(root)
        if node==False:
            return False
        else:
            return True

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int res=0;
    int diameterOfBinaryTree(TreeNode* root) {
        heightofnode(root);
        return res;
    }
    int heightofnode(TreeNode* curr){
        if(curr==nullptr){
            return 0;
        }
            int left=heightofnode(curr->left);
            int right=heightofnode(curr->right);
            res=max(res,left+right);
            return 1+max(left,right);
        
    }
};
